### PR TITLE
Bump minimum SQLite version to 3.8

### PR DESCRIPTION
**Pull request by yahonda**
### Summary

These OS versions have SQLite 3.8 or higher by default.

- macOS 10.10 (Yosemite) or higher
- Ubuntu 14.04 LTS or higher

Raising the minimum version of SQLite 3.8 introduces these changes:

- All of the bundled adapters support `supports_multi_insert?`
- SQLite 3.8 always satisifies `supports_foreign_keys_in_create?` and `supports_partial_index?`
- sqlite adapter can support `alter_table` method for foreign key referenced tables by #32865

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb [hunk 1]
```diff
@@ -48,9 +48,7 @@ def visit_TableDefinition(o)
               statements.concat(o.indexes.map { |column_name, options| index_in_create(o.name, column_name, options) })
             end
 
-            if supports_foreign_keys_in_create?
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb]: This isn't true for the abstract adapter, so I think we still need the conditional for now. (Or, if we're okay to declare that a baseline requirement, it'd need to change there too.)

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb [hunk 2]
```diff
@@ -1062,13 +1062,7 @@ def assume_migrated_upto_version(version, migrations_paths)
           if (duplicate = inserting.detect { |v| inserting.count(v) > 1 })
             raise "Duplicate migration #{duplicate}. Please renumber your migrations to resolve the conflict."
           end
-          if supports_multi_insert?
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: :+1: happy to kill this one

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb [hunk 3]
```diff
@@ -140,7 +144,7 @@ def supports_json?
       end
 
       def supports_multi_insert?
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: I think we can just delete this method and leave the inherited one. And as we're removing the caller, I think we should deprecate the method in the abstract adapter.

File: activerecord/test/cases/schema_dumper_test.rb [hunk 4]
```diff
@@ -192,7 +174,7 @@ def test_schema_dumps_index_columns_in_right_order
 
   def test_schema_dumps_partial_indices
     index_definition = dump_table_schema("companies").split(/\n/).grep(/t\.index.*company_partial_index/).first.strip
-    if current_adapter?(:PostgreSQLAdapter, :SQLite3Adapter) && ActiveRecord::Base.connection.supports_partial_index?
+    if current_adapter?(:PostgreSQLAdapter, :SQLite3Adapter)
```
Comment (matthewd) [inline activerecord/test/cases/schema_dumper_test.rb]: I'm not sure whether this should be using an adapter type check, or the `supports_partial_index?` condition (as we do for index sort order below). Agree it definitely shouldn't be both.

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb [hunk 5]
```diff
@@ -104,6 +104,10 @@ def initialize(connection, logger, connection_options, config)
         @active     = true
         @statements = StatementPool.new(self.class.type_cast_config_to_integer(config[:statement_limit]))
 
+        if sqlite_version < "3.8.0"
+          raise "Your version of SQLite (#{sqlite_version.inspect.match(/(\[\d+\,\ \d+\,\ \d+\])/)[0]}) is too old. Active Record supports SQLite >= 3.8."
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Let's add a `to_s` to AbstractAdapter::Version.

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb [hunk 6]
```diff
@@ -137,6 +137,10 @@ def initialize(version_string)
         def <=>(version_string)
           @version <=> version_string.split(".").map(&:to_i)
         end
+
+        def to_s
+          @version.to_s
```
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb]: `@version.join(".")`

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb [hunk 7]
```diff
@@ -104,6 +104,13 @@ def initialize(connection, logger, connection_options, config)
         @active     = true
         @statements = StatementPool.new(self.class.type_cast_config_to_integer(config[:statement_limit]))
 
+        if sqlite_version < "3.8.0"
+          $stderr.puts <<-EOS.squish
+            "Your version of SQLite (#{sqlite_version.to_s}) is too old. Active Record supports SQLite >= 3.8."
+          EOS
+          exit 1
```
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Is there any reason not to use `raise` unlike other adapters?

https://github.com/rails/rails/blob/ce4d467f7c2fc812e257a87bd4875c1f1f08a981/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb#L56-L58

https://github.com/rails/rails/blob/ce4d467f7c2fc812e257a87bd4875c1f1f08a981/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L220-L222

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb [hunk 8]
```diff
@@ -104,6 +104,13 @@ def initialize(connection, logger, connection_options, config)
         @active     = true
         @statements = StatementPool.new(self.class.type_cast_config_to_integer(config[:statement_limit]))
 
+        if sqlite_version < "3.8.0"
+          $stderr.puts <<-EOS.squish
+            "Your version of SQLite (#{sqlite_version.to_s}) is too old. Active Record supports SQLite >= 3.8."
```
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Unneeded the redundant `to_s` in string interpolation.

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb (hunk 6 above)
Comment (yahonda) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb]: Updated.

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb (hunk 7 above)
Comment (yahonda) [reply activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: I'm thinking about implementing version check like rails/rails#28939 just showing a necessary message for users by omitting stack traces like this.

```ruby
$ sqlite3 --version
3.7.9 2011-11-01 00:52:41 c7c6050ef060877ebe77b41d959e9df13f8c9b5e
```
```ruby
$ bundle exec ruby -w -Itest test/cases/adapters/sqlite3/sqlite3_adapter_test.rb
Using sqlite3
"Your version of SQLite (3.7.9) is too old. Active Record supports SQLite >= 3.8."
$
```
I'd like to get some feedback on this change. I agree to make all bundled adapters behave similarly for the database version check. then I'll revert this change or open another pull request to change mysql2 and posgresql adapter version checks.

Comment (kamipo) [reply activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Usually, a library does/should not terminate a process itself.
If a library may terminate a process itself, it hard to handle that case by users.

Comment (yahonda) [reply activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Understood. Thanks for the comment.

**Pull request merged**